### PR TITLE
Rolling time series aggregation.

### PR DESCRIPTION
**Pull request by ryzhyk**
See individual commits.

### CODE REVIEW

File: src/algebra/mod.rs [hunk 1]
```diff
@@ -210,28 +211,25 @@ where
...
-        + AddAssign
-        + AddAssignByRef
-        + 'static
+impl<T> SemigroupValue for T where
+    T: Clone + Eq + SizeOf + Add<Output = Self> + AddByRef + AddAssign + AddAssignByRef + 'static
 {
 }
 
+/// A type with an associative addition and a zero.
+pub trait MonoidValue: SemigroupValue + HasZero {}
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/algebra/mod.rs:228]: They are both used and serve different purposes.  I'm not sure how to unify them (at least not without rewriting everything).

File: src/operator/time_series/radix_tree/updater.rs [hunk 2]
```diff
@@ -0,0 +1,406 @@
+use super::{ChildPtr, Prefix, RadixTreeCursor, TreeNode, RADIX_BITS};
```
Comment (mbudiu-vmw) [inline src/operator/time_series/radix_tree/updater.rs:1]: perhaps I missed it, but are there tests that compare tree-based aggregation with brute-force aggregation for correctness?

File: src/operator/time_series/radix_tree/updater.rs [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+
+    fn finish(mut self) {
+        while !self.stack.is_empty() {
+            self.pop();
+        }
+
+        // TODO: This can be expensive.  Can we keep `updates` ordered so no sorting is
+        // required?  Currently out-of-order updates are added when creating
+        // intermediate tree nodes (`push_new`).  I can think of two tricks to
+        // reduce the cost of sorting: (1) leave gaps in the updates vector when
```
Comment (mbudiu-vmw) [inline src/operator/time_series/radix_tree/updater.rs:142]: there are some clever algorithms that perform insertion sort using gaps. Rob talked about some of them.

Comment (mbudiu-vmw) [reply src/operator/time_series/radix_tree/updater.rs:142]: I never see lists used in Rust

File: src/operator/time_series/rolling_aggregate.rs [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+            if input_trace_cursor.key_valid() && input_trace_cursor.key() == delta_cursor.key() {
+                debug_assert!(tree_cursor.key_valid());
+                debug_assert_eq!(tree_cursor.key(), delta_cursor.key());
+
+                let mut tree_partition_cursor = PartitionCursor::new(&mut tree_cursor);
+                let mut input_range_cursor =
+                    RangeCursor::new(PartitionCursor::new(&mut input_trace_cursor), ranges);
+
+                // For all affected times, seek them in `input_trace`, compute aggregates using
+                // compute aggregates using radix_tree.
```
Comment (mbudiu-vmw) [inline src/operator/time_series/rolling_aggregate.rs:275]: typo in comment

File: src/operator/time_series/rolling_aggregate.rs [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+        trace::{Batch, BatchReader, Cursor},
+        Circuit, CollectionHandle, DBSPHandle, OrdIndexedZSet, Runtime, Stream,
+    };
+
+    type DataBatch = OrdIndexedZSet<u64, (u64, u64), isize>;
+    type DataStream = Stream<Circuit<()>, DataBatch>;
+    type OutputBatch = OrdIndexedZSet<u64, (u64, (u64, Option<u64>)), isize>;
+    type OutputStream = Stream<Circuit<()>, OutputBatch>;
+
+    fn aggregate_range_slow(batch: &DataBatch, partition: u64, range: Range<u64>) -> Option<u64> {
```
Comment (mbudiu-vmw) [inline src/operator/time_series/rolling_aggregate.rs:341]: I guess this is the brute-force test.

**Review by mbudiu-vmw — APPROVED** [review]
I can't say I can reproduce this, but I have a vague understanding of it.

File: src/nexmark/queries/q18.rs [hunk 6]
```diff
@@ -40,7 +41,7 @@ pub fn q18(input: NexmarkStream) -> Q18Stream {
     });
 
     bids_by_auction_bidder
-        .aggregate::<(), _>(Fold::new(Bid::default(), |top: &mut Bid, val: &Bid, _w| {
+        .aggregate::<(), _>(<Fold<_, UnimplementedSemigroup<_>, _, _>>::new(Bid::default(), |top: &mut Bid, val: &Bid, _w| {
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/nexmark/queries/q18.rs:44]: Aggregators are now parameterized by `Semigroup`, but it is not currently used, so I use placeholder for now.

File: src/circuit/circuit_builder.rs [hunk 7]
```diff
@@ -2430,6 +2509,142 @@ where
...
+    input_stream3: Stream<C, I3>,
+    input_stream4: Stream<C, I4>,
+    output_stream: Stream<C, O>,
+    // `true` if `input_stream1` is an alias to `input_stream2`, `input_stream3` or
+    // `input_stream4`.
+    is_alias1: bool,
+    // `true` if `input_stream2` is an alias to `input_stream3` or `input_stream4`.
+    is_alias2: bool,
+    // `true` if `input_stream3` is an alias to `input_stream4`.
+    is_alias3: bool,
```
[the remark this answers is not in the document]
Comment (Kixiron) [reply src/circuit/circuit_builder.rs:2526]: Two extra bytes doesn't really matter honestly, we could use the `bitflags` crate but I don't know if it's really worth the effort

File: src/operator/aggregate/fold.rs [hunk 8]
```diff
@@ -16,13 +21,14 @@ use std::convert::identity;
 /// * `SF` - step function
 /// * `OF` - output function
 #[derive(Clone)]
-pub struct Fold<A, SF, OF> {
+pub struct Fold<A, S, SF, OF> {
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/aggregate/fold.rs:24]: We should discuss this API. Maybe there's a better way.

File: src/operator/aggregate/fold.rs [hunk 9]
```diff
@@ -34,29 +40,37 @@ impl<A, SF> Fold<A, SF, fn(A) -> A> {
...
     R: MonoidValue,
     A: Clone,
-    SF: Fn(&mut A, &V, R),
-    OF: Fn(A) -> O,
+    SF: Fn(&mut A, &V, R) + Clone,
+    OF: Fn(A) -> O + Clone,
+    S: Semigroup<O> + Clone,
 {
     type Output = O;
+    type Semigroup = S;
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/aggregate/fold.rs:71]: It is used in `radix_tree`.

File: src/operator/time_series/range.rs [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+use crate::trace::Cursor;
+use num::PrimInt;
+use std::{cmp::max, marker::PhantomData};
+
+/// Relative time offset.
+///
+/// Specifies relative time as an offset.
+#[derive(Clone, Eq, PartialEq)]
+pub enum RelOffset<TS> {
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/time_series/range.rs:11]: Timestamps are typically unsigned, so no negative values.

File: src/operator/time_series/range.rs [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    TS: PrimInt,
+{
+    pub fn new(from: RelOffset<TS>, to: RelOffset<TS>) -> Self {
+        Self { from, to }
+    }
+
+    /// Computes relative range of timestamp `ts`.
+    pub fn range_of(&self, ts: &TS) -> Range<TS> {
+        let from = match self.from {
+            RelOffset::Before(off) => ts.saturating_sub(off),
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/time_series/range.rs:36]: We don't want to crash or wrap around if the offset crosses the Unix epoch.

File: src/operator/time_series/radix_tree/mod.rs [hunk 12]
```diff
@@ -0,0 +1,874 @@
...
+//! relative to the event time of the input record.  This may require iterating
+//! over a large volume of data for each event.  Out-of-order inputs make things
+//! worse, as they require re-computing the aggregate multiple times for each
+//! record.  Much of this work is redundant, as the contents of the 3-months
+//! time window for adjacent events is nearly identical, so it should be
+//! possible to reuse most of the computation.  One way to achieve this
+//! is to cover the entire timeline with a radix tree, where each node stores
+//! the value of the aggregate over the time range determined by its bit prefix.
+//!
+//! The following diagram shows a radix tree for a time series with 6-bit
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/time_series/radix_tree/mod.rs:15]: They're not, timestamps must be `PrimInt`.

File: src/operator/time_series/radix_tree/mod.rs [hunk 13]
```diff
@@ -0,0 +1,874 @@
...
+                        assert_eq!(&agg, &child_ptr.child_agg);
+                    } else {
+                        // Validate intermediate node value.
+                        let mut accumulator: Option<A> = None;
+                        for (_, key_agg) in contents
+                            .iter()
+                            .filter(|(&k, _)| child_ptr.child_prefix.contains(k))
+                        {
+                            match accumulator {
+                                None => accumulator = Some(key_agg.clone()),
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/time_series/radix_tree/mod.rs:311]: > does this trick convert any semigroup into a monoid?

yes

> then perhaps you could just require a monoid and have a utility function packaging a semigroup into a monoid.

I tried that, it almost works but not quite.  Problem is that in some cases we represent the neutral value as `None`, while in other cases it is implicitly modeled as missing value.

File: src/operator/time_series/radix_tree/mod.rs [hunk 14]
```diff
@@ -0,0 +1,874 @@
...
+        }
+    }
+}
+
+/// Radix tree node.
+#[derive(Clone, Debug, Default, SizeOf, PartialEq, Eq, PartialOrd, Ord)]
+pub struct TreeNode<TS, A> {
+    /// Array of children.
+    // `Option` doesn't introduce space overhead.
+    children: [Option<ChildPtr<TS, A>>; RADIX],
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/time_series/radix_tree/mod.rs:533]: yes, RADIX slots for children, but some of them may be missing.

File: src/operator/time_series/radix_tree/partitioned_tree_aggregate.rs [hunk 15]
```diff
@@ -0,0 +1,856 @@
...
+
+type OrdPartitionedRadixTree<PK, TS, A, R> = OrdIndexedZSet<PK, (Prefix<TS>, TreeNode<TS, A>), R>;
+type OrdPartitionedRadixTreeStream<PK, TS, A, R> =
+    Stream<Circuit<()>, OrdPartitionedRadixTree<PK, TS, A, R>>;
+
+/// Cursor over partitioned radix tree.
+pub trait PartitionedRadixTreeCursor<'a, PK, TS, A, R>:
+    Cursor<'a, PK, (Prefix<TS>, TreeNode<TS, A>), (), R> + Sized
+{
+    /// Produce a semi-human-readable representation of the partitioned tree
```
[the remark this answers is not in the document]
Comment (ryzhyk) [reply src/operator/time_series/radix_tree/partitioned_tree_aggregate.rs:62]: It'd conflict with `impl Debug for Cursor`.

File: src/operator/time_series/radix_tree/updater.rs (hunk 3 above)
Comment (ryzhyk) [reply src/operator/time_series/radix_tree/updater.rs:142]: > there are some clever algorithms that perform insertion sort using gaps. Rob talked about some of them.

yes, that's the way to do.  I didn't find a good library for it and didn't want to optimize prematurely.

> I never see lists used in Rust

malloc's are expensive.

File: src/operator/time_series/radix_tree/updater.rs (hunk 2 above)
Comment (ryzhyk) [reply src/operator/time_series/radix_tree/updater.rs:1]: Yes, `validate` does this.

**Pull request merged**